### ISSUE TITLE
TTextStyle.RightToLeft in TextRec not working

### ISSUE THREAD

**Issue by almontasser** [bgrabitmap/lazpaint#6]
TBGRABitmap.TextRec result  (the forth line of the image attached) Should be as the TCanvas.TextRec result (the third line of the image attached).

![image](https://cloud.githubusercontent.com/assets/19656179/17942189/a5f55220-6a36-11e6-8063-20c7dfa1d7ea.jpg)

and here is the code I used:
```pascal
procedure TForm1.FormPaint(Sender: TObject);
var
  image: TBGRABitmap;
  TS: TTextStyle;
begin
  image := TBGRABitmap.Create(ClientWidth, ClientHeight);
  image.FontName := Font.Name;
  image.FontHeight := Font.Height;
  TS.Opaque := False;

  TS.RightToLeft := False;
  Canvas.TextRect(Rect(5, 5, 250, 20), 5, 5, 'TCanvas With RightToLeft := False', TS);
  Canvas.TextRect(Rect(250, 5, 500, 20), 250, 5, 'أب+', TS);

  image.TextRect(Rect(5, 20, 250, 35), 5, 20, 'TBGRABitmap With RightToLeft := False', TS, BGRA(0,0,0));
  image.TextRect(Rect(250, 20, 500, 35), 250, 20, 'أب+', TS, BGRA(0,0,0));
  image.Draw(Canvas, 0, 0, False);

  TS.RightToLeft := True;
  Canvas.TextRect(Rect(5, 35, 250, 50), 5, 35, 'TCanvas With RightToLeft := True', TS);
  Canvas.TextRect(Rect(250, 35, 500, 50), 250, 35, 'أب+', TS);

  image.TextRect(Rect(5, 50, 250, 65), 5, 50, 'TBGRABitmap With RightToLeft := True', TS, BGRA(0,0,0));
  image.TextRect(Rect(250, 50, 500, 65), 250, 50, 'أب+', TS, BGRA(0,0,0));
  image.Draw(Canvas, 0, 0, False);

  image.Free;
end;
```

**Comment by lainz** [comment]
Sure it's easy to solve.

In Lazarus trunk there is UTF8ReverseString https://github.com/alrieckert/lazarus/blob/master/components/lazutils/lazutf8.pas#L141

That can be used to reverse the string when that option is present. But warning: That function is not available with Lazarus 1.6 and FPC 3.0. Make sure to keep that in account.

I can't resolve this by myself because I can't track where the real source of TextRect in BGRABitmap is, there are a lot of abstract methods and I'm lost.

**Comment by almontasser** [comment]
That didn't solve it.

The problem not in the order of the letters, the problem is when you mix right to left text with left to right text, like Arabic and English, or Arabic and Numbers, or Arabic and Symbols.

I'know I tried to locate it, and I got stuck.

**Comment by lainz** [comment]
Yes that happened to my the first time I tried to reverse the string. Everything was reversed and not only the part that is not english.

You know where to find the working code in TCanvas?

**Comment by almontasser** [comment]
[https://github.com/alrieckert/lazarus/blob/master/lcl/include/canvas.inc#L1275](url)

```pascal
  if Style.RightToLeft then
    Options := Options or DT_RTLREADING;
  
  ...

    DrawText(DC, pChar(Text), Length(Text), fRect, Options);
```

**Comment by lainz** [comment]
Is strange because reading the BGRABitmap code it uses the Canvas to draw the Text, uses the same functions you're using with Canvas internally.

**Comment by almontasser** [comment]
It uses TextOut
[https://github.com/bgrabitmap/lazpaint/blob/master/bgrabitmap/bgratext.pas#L720](url)

**Comment by lainz** [comment]
You're right, that's the bug

**Comment by lainz** [comment]
I don't know if uses exactly that line, I followed to this, maybe I'm wrong
https://github.com/bgrabitmap/lazpaint/blob/master/bgrabitmap/bgratext.pas#L1220

**Comment by almontasser** [comment]
I gusse It will require a lot of work to support RightToLeft

**Comment by almontasser** [comment]
TCanvas uses ExtUTF8Out, and I was able to make print correctly when I pass it (ETO_RTLREADING=128) in the Options parameter.

https://github.com/alrieckert/lazarus/blob/master/lcl/include/canvas.inc#L1351

**Comment by circular17** [comment]
Indeed that would be work to handle RightToLeft. In the meantime, you can use the Canvas property of BGRABitmap. It works ok as long as images have an opaque background.
